test/common/network/udp_listener_impl_batch_writer_test.cc

```diff
@@ -0,0 +1,300 @@
+ *      Internal Buffer.
+ *    - A buffer longer than the initial buffer should not get appended to the
+ *      Internal Buffer.
+ */
+TEST_P(UdpListenerImplBatchWriterTest, WriteBlocked) {
+  // Quic Mock Objects
+  quic::test::MockQuicSyscallWrapper os_sys_calls;
+  quic::ScopedGlobalSyscallWrapperOverride os_calls(&os_sys_calls);
+
+  // The initial payload to be buffered
+  std::string initial_payload("length7");
+  Buffer::InstancePtr initial_buffer(new Buffer::OwnedImpl());
+  initial_buffer->add(initial_payload);
+  UdpSendData initial_send_data{send_to_addr_->ip(), *server_socket_->localAddress(),
+                                *initial_buffer};
+
+  // Get initial value of total_bytes_sent
+  uint64_t total_bytes_sent =
+      listener_config_.listenerScope().counterFromString("total_bytes_sent").value();
+
+  // Possible followup payloads to be sent after the initial payload
+  absl::FixedArray<std::string> followup_payloads{"length<7", "len<7"};
+
+  for (const auto& followup_payload : followup_payloads) {
+    std::string internal_buffer("");
+
+    // First have initial payload added to the udp_packet_writer's internal buffer.
+    auto send_result = listener_->send(initial_send_data);
+    internal_buffer.append(initial_payload);
+    EXPECT_TRUE(send_result.ok());
+    EXPECT_FALSE(listener_->udpPacketWriter()->isWriteBlocked());
+    EXPECT_EQ(listener_config_.listenerScope()
+                  .gaugeFromString("internal_buffer_size", Stats::Gauge::ImportMode::NeverImport)
+                  .value(),
+              initial_payload.length());
+    EXPECT_EQ(listener_config_.listenerScope()
+                  .gaugeFromString("front_buffered_pkt_size", Stats::Gauge::ImportMode::NeverImport)
+                  .value(),
+              initial_payload.length());
+    EXPECT_EQ(listener_config_.listenerScope().counterFromString("total_bytes_sent").value(),
+              total_bytes_sent);
+
+    // Mock the socket to be write blocked on sendmsg syscall
+    EXPECT_CALL(os_sys_calls, Sendmsg(_, _, _))
+        .WillOnce(Invoke([](int /*sockfd*/, const msghdr* /*msg*/, int /*flags*/) {
+          errno = EWOULDBLOCK;
+          return -1;
+        }));
+
+    // Now send the followup payload
+    Buffer::InstancePtr followup_buffer(new Buffer::OwnedImpl());
+    followup_buffer->add(followup_payload);
+    UdpSendData followup_send_data{send_to_addr_->ip(), *server_socket_->localAddress(),
+                                   *followup_buffer};
+    send_result = listener_->send(followup_send_data);
+
+    // The followup payload should only get buffered if it is shorter than initial payload
+    if (followup_payload.length() < initial_payload.length()) {
+      EXPECT_TRUE(send_result.ok());
+      // TODO(yugant): This flag should be true here, but currently it is incorrectly set in
+      // quiche code. Change this to True, once this issue is fixed in quiche.
+      EXPECT_FALSE(listener_->udpPacketWriter()->isWriteBlocked());
```

Comment:
I don't think this is a bug in QUICHE. We only set writer blocked upon WRITE_STATUS_BLOCKED. If it's WRITE_STATUS_BLOCKED_DATA_BUFFERED, from callers perspective the current write is still successful because data are taken by the writer. And the caller will know the socket is actually buffered upon next write.

Reply:
I had put in this comment here after discussing this with Bin. Based on our discussions, the logic was that the `sendmsg` syscall failed here because of the socket being blocked. Therefore it made sense to mark the `write_blocked` flag, as any subsequent `sendmsg` on the socket will also fail unless the writer is unblocked. 
This may avoid one extra failed attempt to write on a blocked socket. (For which we will then receive `WRITE_STATUS_BLOCKED`).

That being said, I understand your reasoning for this. Since we are changing the way the quic_results are interpreted, we can can consider `WRITE_STATUS_BLOCKED_DATA_BUFFERED` as a success case as well. 
I have updated the test here to expect the same.